kernel/work_q.c

```diff
@@ -116,18 +114,18 @@ int k_delayed_work_cancel(struct k_delayed_work *work)
 {
 	int key = irq_lock();
 
-	if (k_work_pending(&work->work)) {
-		irq_unlock(key);
-		return -EINPROGRESS;
-	}
-
 	if (!work->work_q) {
 		irq_unlock(key);
 		return -EINVAL;
 	}
 
-	/* Abort timeout, if it has expired this will do nothing */
-	_abort_timeout(&work->timeout);
+	if (k_work_pending(&work->work)) {
+		/* Remove from the queue if already submitted */
+		k_queue_find_and_remove(&work->work_q->queue, work);
```

Comment:
Perhaps we should have a mutex/semaphore for work_q so we don't have to use irq_lock in order to cancel.